Add travis configuration

leandro-lucarella-sociomantic opened the pull request.

leandro-lucarella-sociomantic committed:
Add travis configuration

mihails-strasuns-sociomantic commented:
FYI: this is how one can force old version of dmd (or set of versions) - https://github.com/rejectedsoftware/vibe.d/blob/master/.travis.yml#L8-L17

leandro-lucarella-sociomantic commented:
Yeah, but I don't want to use travis facilities, as you can't pick which underlying distribution to use. We need to test with particular Ubuntu versions, because we have a lot of external library dependencies.

leandro-lucarella-sociomantic commented:
Almost green now, now I'm getting some [D2 compile errors](https://travis-ci.org/sociomantic-tsunami/ocean/builds/186162198#L2867):

```
./src/ocean/util/container/cache/CachingStructLoader.d(104): Error: template instance ocean.util.container.cache.ExpiringCache.ExpiringCache!16LU error instantiating
./src/ocean/util/container/cache/CachingStructLoader_test.d(75):        instantiated from here: CachingStructLoader!(Trivial)
./src/ocean/util/container/cache/CachingStructLoader_test.d(142):        instantiated from here: TestCache!(Trivial)
./src/ocean/util/container/cache/CachingStructLoader.d(49): Error: class ocean.util.container.cache.CachingStructLoader.CachingStructLoader!(Trivial).CachingStructLoader destructors, postblits and invariants are not allowed in overlapping fields fiber_local_copy and cache_
./src/ocean/util/container/cache/CachingStructLoader_test.d(79): Error: no property 'Cache' for type 'object.Object'
./src/ocean/util/container/cache/CachingStructLoader_test.d(88): Error: function ocean.util.container.cache.CachingStructLoader_test.TestCache!(Trivial).TestCache.Cache.now does not override any function
./src/ocean/util/container/cache/CachingStructLoader_test.d(118): Error: function ocean.util.container.cache.CachingStructLoader_test.TestCache!(Trivial).TestCache.getData does not override any function
```

Any ideas @mihails-strasuns-sociomantic ?

leandro-lucarella-sociomantic commented:
This is with:
```
Setting up dmd-bin (2.071.2-0)
```

leandro-lucarella-sociomantic commented:
Removing the `[WIP] ` from the subject because all the operational part is ready, we just need to figure out compilation errors (all the D1 build passes for example).

mihails-strasuns-sociomantic commented:
> Setting up dmd-bin (2.071.2-0)

Well, this is your problem right here :) dmd-transitional is based on 2.070.2

leandro-lucarella-sociomantic commented:
Now we have a unit test error. Almost there!
```
   run build-d2/production/tmp/allunittests
core.exception.AssertError@./src/ocean/core/Exception.d(258): unittest failure
----------------
??:? _d_unittest [0x10c2130]
??:? void ocean.core.Exception.__unittest_fail(int) [0xb6b2b4]
./src/ocean/core/Exception.d:258 void ocean.core.Exception.__unittestL253_4() [0xb6a1fd]
??:? void ocean.core.Exception.__modtest() [0xb6b22d]
./src/ocean/core/UnitTestRunner.d:628 ocean.core.UnitTestRunner.UnitTestRunner.Result ocean.core.UnitTestRunner.UnitTestRunner.timedTest(object.ModuleInfo*, out core.sys.posix.sys.time.timeval, ref char[]) [0xe8accf]
./src/ocean/core/UnitTestRunner.d:243 _D5ocean4core14UnitTestRunner14UnitTestRunner3runMFZ14__foreachbody1MFPS6object10ModuleInfoZi [0xe88b99]
??:? int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)).__lambda2(immutable(object.ModuleInfo*)) [0x10c148f]
??:? int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))).__foreachbody2(ref rt.sections_elf_shared.DSO) [0x10cadbe]
??:? int rt.sections_elf_shared.DSO.opApply(scope int delegate(ref rt.sections_elf_shared.DSO)) [0x10cb289]
??:? int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))) [0x10cad4f]
??:? int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)) [0x10c146b]
./src/ocean/core/UnitTestRunner.d:188 int ocean.core.UnitTestRunner.UnitTestRunner.run() [0xe88671]
./src/ocean/core/UnitTestRunner.d:923 _Dmain [0xe8bae6]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x10c645e]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x10c639c]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x10c641a]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x10c639c]
??:? _d_run_main [0x10c62f9]
??:? main [0xe8c1b5]
??:? __libc_start_main [0x76810f44]
```

leandro-lucarella-sociomantic commented:
> Well, this is your problem right here :) dmd-transitional is based on 2.070.2

BTW, hurray for compiler stability!

nemanja-boric-sociomantic commented:
There are `-transition=import -transition=checkpoimports` switches which should serve for deprecation cycle.

leandro-lucarella-sociomantic commented:
@mihails-strasuns-sociomantic any ideas about the unit test failure?

mathias-lang-sociomantic commented:
The runtime doesn't `Throwable.message`.
As a result, `getMsg` will use [this branch](https://github.com/sociomantic-tsunami/ocean/blob/508796b5183f9c29fcdce57e98a1012c7311085b/src/ocean/transition.d#L881-L886).
And `toString` returns the whole stack trace.

mihails-strasuns-sociomantic commented:
Yes, @mathias-lang-sociomantic is correct, failures are because of `getMsg` mismatch. We could probably hack this but I am not sure what is your goal here - there is no way we could use upstream DMD release for actual ocean development, why bother with setting it up in CI at all?

MartinNowak commented:
> Yeah, but I don't want to use travis facilities, as you can't pick which underlying distribution to use. We need to test with particular Ubuntu versions, because we have a lot of external library dependencies.

Maybe you can, see [The Trusty beta Build Environment - Travis CI](https://docs.travis-ci.com/user/trusty-ci-environment/). It doesn't replace proper testing of your production environment though.
Maybe use docker base images? Those can be run on Travis-CI AFAIK.

leandro-lucarella-sociomantic commented:
> Maybe use docker base images? Those can be run on Travis-CI AFAIK.

We are using docker already, the problem is not that, the problem is we are using a custom dmd compiler, internally called `dmd-transitional`, to overcome a couple of issues we have.

@mihails-strasuns-sociomantic my idea was to have at least something working publicly, without having to publish our `dmd-transitional`, but if we are moving the development to this repo, I guess we'll have to.

leandro-lucarella-sociomantic commented:
We'd really **really** want to get rid of `dmd-transitional` instead of publishing it, that would make D's ecosystem even worse, but to do this we need a solution for `Throwable.message`.

This change was [merged a while ago](https://github.com/dlang/druntime/pull/1445) and then [reverted](https://github.com/MartinNowak/druntime/commit/80e5059bed9235f2cdebea2ca5366898b36ed506).

This is the only thing that prevents Sociomantic from using the official DMD compiler, and now is a blocker to move our development of ocean into the wild.

@MartinNowak how is that rc string going, is there anything we can do to speed this up to reach a final solution ASAP?

leandro-lucarella-sociomantic commented:
I was just going through the old issue and I realized it's been 1 year and 13 days since I wrote this: https://github.com/dlang/druntime/pull/1445#issuecomment-164535171 and we are still at the exact same place. :cry:

mihails-strasuns-sociomantic commented:
@leandro-lucarella-sociomantic so, what are you going to do with this PR considering last information? Settle with partial testing for now or block on public release of `dmd-transitional`?

leandro-lucarella-sociomantic commented:
For now wait for an answer from upstream on the issue that prevent us from using the upstream compiler.

leandro-lucarella-sociomantic committed:
travis: Use C language

By default ruby is used, which makes no sense. Using D is also not a good
idea because we don't use any features from the image and it's in beta, so
using a lightweight C image seems reasonable.

leandro-lucarella-sociomantic committed:
travis: Revamp docker handling and push to registry

A few assorted improvements to travis scripts. First split into separated
scripts the before_install and script commands. Then use the new
sociomantictsunami Docker Hub organization. Finally try to pull the image
from the registry, so we can use it as a cache, and try to push it back
if there are credentials available.

Also use $TRAVIS_REPO_SLUG to get the docker image name, so scripts are
more reusable.

leandro-lucarella-sociomantic committed:
docker: Move scripts used by CI to ci

Mixing stuff run by docker's image creation and stuff that run inside
the container is a bad idea, because any changes in the files copied to
the docker daemon, can potentially invalidate the image cache triggering
an unnecessary rebuild.

Because of this is better to have scripts used to build the docker image
separated from the scripts used to actually run the tests in a CI
system.

This commit also moves the Dockerfile to the top-level directory to make
it easier to just run `docker build .` (for which a `.dockerignore` file
is added, so only stuff in `docker/` is copied to the docker daemon).

leandro-lucarella-sociomantic committed:
docker: Build dmd-transitional compiler

To be able to test Ocean we need to build a dmd-transitional compiler
which includes some changes to upstream DMD that are still under
discussion.

See:
* https://github.com/dlang/DIPs/pull/55
* https://github.com/dlang/druntime/pull/1445
* https://github.com/MartinNowak/druntime/commit/80e5059bed9235f2cdebea2ca5366898b36ed506

leandro-lucarella-sociomantic committed:
Allow overriding of DC (D compiler) variable

The DC variable in the Build.mak file was being set unconditionally,
which complicates using a different compiler to build.

leandro-lucarella-sociomantic commented:
:green_heart: :sunglasses:

leandro-lucarella-sociomantic committed:
docker: Add labels with dependency versions

leandro-lucarella-sociomantic commented:
@mihails-strasuns-sociomantic @nemanja-boric-sociomantic it would be useful if you can have a quick look at this so it can be merged soon(ish).

@gavin-norman-sociomantic also FYI this should sort of unblock v3 release when merged (although we might need to talk about how to do the transition to move the development to the public repo)

leandro-lucarella-sociomantic commented:
I will actually rebase this to v2.4.x so we can then merge to newer releases so all branches get tested.

leandro-lucarella-sociomantic committed:
travis: Always pull the latest image when building

nemanja-boric-sociomantic commented on docker/base.sh:
```diff
@@ -0,0 +1,44 @@
+
+# Install packages that depends on the distro
+ruby=ruby1.9.1-dev
+test $release = xenial &&
```
since there's `FROM ubuntu:trusty`, is this necessary?

nemanja-boric-sociomantic replied:
well, even not, it's good to have it.

leandro-lucarella-sociomantic replied:
Not now, but the hardcoded trusty build is just a simplification. In a near future I would want to build for both trusty and xenial (also, this distinction here is not even necessary, there is a `ruby-dev` package that works for both :grin: )

nemanja-boric-sociomantic commented on docker/base.sh:
```diff
@@ -0,0 +1,44 @@
+# Update the whole system
+apt-get -y dist-upgrade
+
+# fpm installation is release-dependant
```
is this comment valid? I don't see anything release-dependent here.

leandro-lucarella-sociomantic replied:
Yes, but maybe it's pointless at this place.

nemanja-boric-sociomantic commented on docker/ci.sh:
```diff
@@ -0,0 +1,28 @@
+    xmlfile="$BUILD_DIR_NAME/$F/tmp/unittests.xml"
+
+    if_d 2 \
+        make -r d2conv
```
nice :+1:

nemanja-boric-sociomantic commented on docker/dmd1.sh:
```diff
@@ -0,0 +1,57 @@
+    vendor = 'Sociomantic Labs GmbH',
+    license = 'GPLv2',
+    depends = ['gcc'] + FUN.autodeps('src/dmd'),
+    deb_recommends = 'dmd1-rt',
```
`libtangort-dmd-dev`?

leandro-lucarella-sociomantic replied:
nope, `dmd1-rt` is a virtual package and `libtangort-dmd-dev` [provides it](https://github.com/sociomantic-tsunami/tangort/blob/v1.x.x/pkg/libtangort-dmd-dev.pkg#L11). Is a more general dependency (not that it was really necessary, but it works :).

leandro-lucarella-sociomantic commented:
@nemanja-boric-sociomantic all the scripts to build the docker image will be moved elsewhere in a near future, and they will be further polished. I don't want to spend much more time on them in this PR as it is still a temporary solution.

nemanja-boric-sociomantic commented on docker/travis.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+set -xe
+
+git submodule foreach --recursive git submodule deinit --force --all
```
what about a comment here?

leandro-lucarella-sociomantic replied:
Could be informative :)

leandro-lucarella-sociomantic replied:
Added comment.

nemanja-boric-sociomantic commented:
> @nemanja-boric-sociomantic all the scripts to build the docker image will be moved elsewhere in a near future, and they will be further polished. I don't want to spend much more time on them in this PR as it is still a temporary solution.

Ok, I'll pay less attention to it (it does not look bad, btw).

leandro-lucarella-sociomantic committed:
travis: Add comment about submodules deinit

nemanja-boric-sociomantic commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+*
+!docker/
```
does gitignore need to ignore this file?

leandro-lucarella-sociomantic replied:
That's `.dockerignore`, not `.gitignore` :)

nemanja-boric-sociomantic replied:
Yes, what I meant is does `.gitignore` needs `.dockerignore` now? :-)

leandro-lucarella-sociomantic replied:
I switched from `docker build -f docker/Dockerfile docker` to `docker build .` because it requires less special treatment. When using Jenkins model definition plug-in it didn't support specifying a custom directory for example, which made docker pass to the docker daemon the full repository (which is a lot of copying). I think is better for the project to be more self-sufficient. Any casual user that doesn't know about the details on how docker works would be much safer if the obvious way to build an image for the project "just works" without needing to use special rules to build it.

leandro-lucarella-sociomantic replied:
So basically the `.dockerignore` file say:
1. `*`: ignore everything
2. `!docker/` except for the `docker` directory

nemanja-boric-sociomantic replied:
Yeah, I got just confused for a moment! I wanted to ignore this file, but this commit actually adds it to git, that's why I said "Oh, nevermind" (seems that that comment is gone now!)

nemanja-boric-sociomantic replied:
All good! Sorry for noise!

nemanja-boric-sociomantic commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+*
+!docker/
```
Oh, nevermind...

nemanja-boric-sociomantic commented on .dockerignore:
```diff
@@ -1,2 +1,2 @@
 *
-!ci/
+!docker/
```
Surely this commit can just be removed, along with the same commit that renamed this stuff in the first place? :-)

nemanja-boric-sociomantic commented on .dockerignore:
```diff
@@ -1,2 +1,2 @@
 *
-!ci/
+!docker/
```
or can at least be squashed with the one that moves `docker` -> `ci`

leandro-lucarella-sociomantic replied:
I could see if I can rearrange commits to make it a bit cleaner, but it was a very long-term work so at some point I just stopped caring. I could also squash everything in one commit I guess. I'll have a look and see what's the best I can make with the biggest outcome/effort ratio.

nemanja-boric-sociomantic replied:
fair enough, don't bother if it would cause too much problems.

nemanja-boric-sociomantic reviewed (approved):
Modulo small nitpicks which I don't find very important, great job!

leandro-lucarella-sociomantic committed:
Use https for submodule URLs

This is necessary to allow unauthenticated fetching.

leandro-lucarella-sociomantic commented:
I just pushed a small revamp on the commits, I mostly squashed the moving of docker -> ci and then back to docker. The rest of the commits I think provide some useful information as they are.

leandro-lucarella-sociomantic commented:
I'll auto-merge tomorrow if I don't find any complains :)

mihails-strasuns commented:
I think some of patches for `dmd-transitional` can be simplified or even removed but I'd better experiment with in in a separate PR after we get CI coverage working.

mihails-strasuns-sociomantic reviewed (approved):
I will like it more when most of the docker stuff gets moved into own repo but really great job nevertheless :)

mihails-strasuns-sociomantic commented:
How do you intend to handle branch merges during migration of main development to public repo?

leandro-lucarella-sociomantic commented:
> How do you intend to handle branch merges during migration of main development to public repo?

Once we flip the switch, same as we did internally.
